hal/mpu/mbed_mpu_v7m.c

```diff
@@ -0,0 +1,156 @@
+     * 0x40000000 - 0x5FFFFFFF     Peripheral      No
+     * 0x60000000 - 0x7FFFFFFF     RAM             Yes                          Execute disabled
+     * 0x80000000 - 0x9FFFFFFF     RAM             Yes                          Execute disabled
+     * 0xA0000000 - 0xBFFFFFFF     Device          No
+     * 0xC0000000 - 0xDFFFFFFF     Device          No
+     * 0xE0000000 - 0xFFFFFFFF     System          No
+     */
+
+    if (regions >= 3) {
+        // Select region 0 and used it for the WT rom region
+        // - RAM 0x10000000 - 0x1FFFFFFF
+        MPU->RNR = 0;
+        // Set address to 0
+        MPU->RBAR = 0;
+        // Configure and enable region
+        MPU->RASR =
+            MPU_RASR_ENABLE_Msk |
+            ARM_MPU_RASR(
+                1,                          // DisableExec
+                ARM_MPU_AP_FULL,            // AccessPermission
+                0,                          // TypeExtField
+                0,                          // IsShareable
+                1,                          // IsCacheable
+                0,                          // IsBufferable
+                                            // SubRegionDisable
+                (1 << 0) |     // Disable Sub-region
+                (1 << 1) |     // Disable Sub-region
+                (1 << 2) |     // Disable Sub-region
+                (1 << 3) |     // Disable Sub-region
+                (0 << 4) |     // Enable Sub-region RAM 0x10000000 - 0x1FFFFFFF
+                (0 << 5) |
+                (0 << 6) |
+                (0 << 7),
+                ARM_MPU_REGION_SIZE_512MB   // Size
+            );
+
+        // Select region 1 and used it for WBWA ram regions
+        // - SRAM 0x20000000 - 0x3FFFFFFF
+        // - RAM  0x60000000 - 0x7FFFFFFF
+        MPU->RNR = 1;
+        // Set address to 0
+        MPU->RBAR = 0;
+        // Configure and enable region
+        MPU->RASR =
+            MPU_RASR_ENABLE_Msk |
+            ARM_MPU_RASR(
+                1,                          // DisableExec
+                ARM_MPU_AP_FULL,            // AccessPermission
+                1,                          // TypeExtField
+                0,                          // IsShareable
+                1,                          // IsCacheable
+                1,                          // IsBufferable
+                                            // SubRegionDisable
+                (1 << 0) |     // Disable Sub-region
+                (0 << 1) |     // Enable Sub-region SRAM 0x20000000 - 0x3FFFFFFF
+                (1 << 2) |     // Disable Sub-region
+                (0 << 3) |     // Enable Sub-region RAM 0x60000000 - 0x7FFFFFFF
+                (1 << 4) |     // Disable Sub-region
+                (1 << 5) |     // Disable Sub-region
+                (1 << 6) |     // Disable Sub-region
+                (1 << 7),      // Disable Sub-region
+                ARM_MPU_REGION_SIZE_4GB     // Size
```

Comment:
The handling of `10000000` feels a fudge to me - either `1` is ROM, in which case it should be treated as `0` (execute, +read-only please), or it's RAM, in which case it should be treated as `2` (WBWA, XN). Can't we have a flag to indicate which mode we want for each platform?

At the minute marking the ROM read-only would require 3 regions because you've got 3 different types of behaviour (X WT, XN WT, XN WBWA). Adding a flag would let us reduce it to (X RO, XN WBWA), and if this region is reduced to 2GB, have all the necessary control.

Reply:
Some targets such as the [LPC407X](https://github.com/ARMmbed/mbed-os/blob/cb4924151e85f88867445c1f4eb39a05ed6f3af0/targets/TARGET_NXP/TARGET_LPC408X/device/TOOLCHAIN_ARM_STD/LPC407X_8X.sct#L6-L15)n have both ROM and RAM in the code region (0x00000000 - 0x1FFFFFFF ). Because of this the extra region is required.

Reply:
That's certainly necessary with the current implementation that turns the MPU off, rather than just flipping the XN control, but we do in principle now have scope to have a "better" base configuration than emulating the MPU-off default.

But I guess we don't have to do it now; this PR does leave the door open to further refinement - it's only really specifying the XN behaviour. The current implementation would directly conflict with an "Ethernet+cache" solution, but it could be reworked as part of trying to integrate that without breaking the APIs defined here.

Reply:
Even without the MPU-off default, changing the cache settings from their default could change devices behavior from what they are now. This is why I kept the cache settings the same. This is unrelated to the MPU-off default.